spec/p2p/p2p.qnt

```diff
@@ -0,0 +1,577 @@
+                    config.get(n).Peers().exists(p => c.address == p)))
+
+    //  if and address appears in connections and peers they should agree on the direction
+    // TODO: might be broken
+    val invDirectionsOut = modeledNodes.forall(n =>
+        config.get(n).Peers().forall(p => 
+        config.get(n).connections.forall(c => (c.address == p and c.state == "outgoing") 
+                                        implies config.get(n).peerState.get(p) == "outbound")))
+
+    // For nostalgic reasons and TLA+ lovers...
+    val invTypeOK = modeledNodes.forall(n =>
+        config.get(n).Peers().forall(p => 
+        config.get(n).connections.forall(c =>
+            connectionStates.contains(c.state) and
+            peerStates.contains(config.get(n).peerState.get(p)))))
+
+    // // for all modeled nodes y and all its peers x, some peerStates imply some connection state
+    // val invConnectionsNotOpen = modeledNodes.forall(y => config.get(y).Peers().forall(x => 
+    //     (config.get(y).peerState.get(x) == "dialing" or config.get(y).peerState.get(x) == "reconnecting") implies 
+    //         config.get(y).connectionState.get(x) != "open"))
+
+    // val invConnectionsNotInit = modeledNodes.forall(y => config.get(y).Peers().forall(x => 
+    //     (config.get(y).peerState.get(x) == "inbound" or config.get(y).peerState.get(x) == "outbound") implies 
+    //         config.get(y).connectionState.get(x) != "init"))
+
+    // val invConnectionsOpen = modeledNodes.forall(y => config.get(y).activePeers().forall(x => 
+    //     config.get(y).peerState.get(x) == "outbound" or config.get(y).peerState.get(x) == "inbound"))
+
+
+
+/* ****************************************************************************
+ * Actions
+ * How the above functions can be invoked (e.g. by which node)
+ * ************************************************************************* */
+
+  
+/* 
+ * Future syntactic sugar.
+ */
+    // helps for better readability. 
+    def require(pred:bool):bool = pred
+
+/*
+ * Useful actions to simplify spec
+ */ 
+
+    action allUnchanged = all {
+        config' = config,
+        Environment' = Environment,
+        OperatorAddressBook' = OperatorAddressBook,
+        requestingNodes' = requestingNodes,
+    }    
+
+    // Framerule
+    action environmentUnchanged = all {
+        Environment' = Environment,
+        OperatorAddressBook' = OperatorAddressBook,
+        requestingNodes' = requestingNodes,
+    }    
+
+    
+
+/*
+ * Local actions
+ */
+
+   // node dials a. This starts lower lever computations. 
+    action actionDialPeer (node: Address, a:Address):bool = all {
+        config' = config.set(node, config.get(node).dialPeer(a)),
+        environmentUnchanged,
+    }
+
+    action actionTransportAccept (node: Address, a:Address):bool = all {
+        config' = config.set(node, config.get(node).transportAccept(a)),
+        environmentUnchanged,
+    }
```

Comment:
if (config.get(node) == config.get(node).transportAccept(a)), that is, if the node doesn't accept connection, then record that in the rawConnection. 
Then having a failed rawConnection should enable dialpeerFailed on the other side (if the other side is  a modeledNode